src/software/ai/passing/pass_evaluation.tpp

```diff
@@ -0,0 +1,84 @@
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+
+#include "software/ai/passing/field_pitch_division.h"
+#include "software/ai/passing/pass.h"
+#include "software/ai/passing/pass_evaluation.h"
+#include "software/ai/passing/pass_with_rating.h"
+#include "software/geom/point.h"
+#include "software/time/timestamp.h"
+
+template <class ZoneEnum>
+PassEvaluation<ZoneEnum>::PassEvaluation(
+    std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division,
+    std::unordered_map<ZoneEnum, PassWithRating> best_pass_in_zones,
+    std::shared_ptr<const PassingConfig> passing_config, Timestamp timestamp)
+    : pitch_division_(pitch_division),
+      best_pass_in_zones_(best_pass_in_zones),
+      passing_config_(passing_config),
+      timestamp_(timestamp)
+{
+}
+
+template <class ZoneEnum>
+PassWithRating PassEvaluation<ZoneEnum>::getBestPassOnField() const
+{
+    auto best_pass =
+        std::max_element(best_pass_in_zones_.begin(), best_pass_in_zones_.end(),
+                         [](const std::pair<ZoneEnum, PassWithRating>& p1,
+                            const std::pair<ZoneEnum, PassWithRating>& p2) {
+                             return p1.second.rating < p2.second.rating;
+                         });
+    return best_pass->second;
+}
+
+template <class ZoneEnum>
+PassWithRating PassEvaluation<ZoneEnum>::getBestPassInZones(
+    const std::unordered_set<ZoneEnum>& zone_ids) const
+{
+    if (zone_ids.size() == 0)
+    {
+        throw std::invalid_argument("no zone_ids provided");
+    }
+
+    auto compare_pass_rating = [this](ZoneEnum zone_a, ZoneEnum zone_b) {
+        return best_pass_in_zones_.at(zone_a).rating <
+               best_pass_in_zones_.at(zone_b).rating;
+    };
+
+    ZoneEnum best_zone =
+        *std::max_element(zone_ids.begin(), zone_ids.end(), compare_pass_rating);
+
+    return best_pass_in_zones_.at(best_zone);
+}
+
+template <class ZoneEnum>
+std::shared_ptr<const FieldPitchDivision<ZoneEnum>>
+PassEvaluation<ZoneEnum>::getFieldPitchDivsion() const
+{
+    return pitch_division_;
+}
+
+template <class ZoneEnum>
+Timestamp PassEvaluation<ZoneEnum>::getEvaluationTime() const
+{
+    return timestamp_;
+}
+
+template <class ZoneEnum>
+std::vector<ZoneEnum> PassEvaluation<ZoneEnum>::rankZonesForCherryPicking(
+    const World& world, const Point& position) const
+{
+    std::vector<ZoneEnum> cherry_pick_zones = pitch_division_->getAllZoneIds();
+
+    std::sort(
```

Comment:
Is cherry picking what's really happening here? Cherry picking implies looking for a shot on net iinm and I'm not sure that rateZone considers shots on net.

Reply:
yeah this might just need to be re-named to something else

the purpose this serves is to support robots that have possession of the ball, allowing them to pass if needed while making progress up the field towards the enemy net.

from the sports definition: Cherry picking, in basketball and certain other sports, refers to play where one player (the cherry picker) does not play defense with the rest of the team but remains near the opponents' goal.

So it definitely doesn't line up with that entirely, even if rateZone considers shot on net, because It seems like the definition of cherry picking limits us to "being near" the enemy goal.